debrsign.1
.TH DEBRSIGN 1 "Debian Utilities" "DEBIAN" \" -*- nroff -*-
.SH NAME
debrsign \- remotely sign a Debian changes and dsc file pair using SSH
.SH SYNOPSIS
\fBdebrsign\fR [options] [\fIuser@\fR]\fIremotehost\fR
[\fIchanges-file\fR|\fIdsc-file\fR]
.SH DESCRIPTION
\fBdebrsign\fR takes either an unsigned \fI.dsc\fR file or an
unsigned \fI.changes\fR file and the associated unsigned \fI.dsc\fR
file (found by replacing the architecture name and \fI.changes\fR
by \fI.dsc\fR) if it appears in the \fI.changes\fR file and signs them
by copying them to the remote machine using \fBssh\fR(1) and remotely
running \fBdebsign\fR(1) on that machine.  All options not listed
below are passed to the \fBdebsign\fR program on the remote machine.
.PP
If a \fI.changes\fR or \fI.dsc\fR file is specified, it is signed,
otherwise, \fIdebian/changelog\fR is parsed to determine the name of
the \fI.changes\fR file to look for in the parent directory.
.PP
This utility is useful if a developer must build a package on one
machine where it is unsafe to sign it; they need then only transfer
the small \fI.dsc\fR and \fI.changes\fR files to a safe machine and
then use the \fBdebsign\fR program to sign them before
transferring them back.  This program automates this process.
.PP
To do it the other way round, that is to connect to an unsafe machine
to download the \fI.dsc\fR and \fI.changes\fR files, to sign them
locally and then to transfer them back, see the \fBdebsign\fR(1)
program, which can do this task.
.SH OPTIONS
.TP
\fB\-S\fR
Look for a source-only \fI.changes\fR file instead of a binary-build
changes file.
.TP
\fB\-a\fIdebian-architecture\fR, \fB\-t\fIGNU-system-type\fR
See \fBdpkg-architecture\fR(1) for a description of these options.
They affect the search for the \fI.changes\fR file.  They are provided
to mimic the behaviour of \fBdpkg-buildpackage\fR when determining the
name of the \fI.changes\fR file.
.TP
\fB\-\-help\fR, \fB\-\-version\fR
Show help message and version information respectively.
.TP
\fBOther options\fR
All other options are passed on to \fBdebuild\fR on the remote
machine.
.SH "SEE ALSO"
.BR debsign (1),
.BR dpkg-architecture (1)
and
.BR ssh (1).
.SH AUTHOR
This program was written by Julian Gilbey <jdg@debian.org> and is
copyright under the GPL, version 2 or later.


debian/NEWS
devscripts (2.10.54) UNRELEASED; urgency=low

    manpage-alert now, as originally intended, prints out the full path to the
    binary which is lacking a manpage instead of simply the binary name.

    Before:
      No manual entry for somebinary

    After:
      No manual entry for /usr/bin/somebinary

 -- James Vega <jamessan@debian.org>  Thu, 06 Aug 2009 18:03:41 -0400

devscripts (2.10.24) unstable; urgency=low

    Starting with devscripts 2.10.21, debchange attempts to
    automatically determine whether an NMU changelog should be
    started, by comparing DEBEMAIL and DEBFULLNAME to the
    package's Maintainer and Uploaders fields.

    This behaviour may now be disabled using the DEBCHANGE_AUTO_NMU
    configuration file variable.

 -- Adam D. Barratt <adam@adam-barratt.org.uk>  Mon, 07 Apr 2008 14:38:08 +0100

devscripts (2.10.20) unstable; urgency=low

    devscripts 2.10.19 modified debcommit's --release option to use the
    uncommitted changes (if any) in the changelog to derive the commit
    message.

    This behaviour is now optional and can be enabled using the -R /
    --release-use-changelog command line option or by setting the
    DEBCOMMIT_RELEASE_USE_CHANGELOG configuration option to "yes".

    The default is to use the previous (i.e. pre-2.10.19) behaviour.

 -- Adam D. Barratt <adam@adam-barratt.org.uk>  Wed, 19 Mar 2008 20:31:20 +0000

devscripts (2.10.19) unstable; urgency=low

    In devscripts 2.10.17, dget was modified to unpack source packages by
    default.  The new -d switch or the DGET_UNPACK configuration file option
    may be used to indicate that the unpacking should not occur.

    As linda has been removed from both unstable and testing, support for
    it has been removed from debuild and cvs-debuild.

 -- Adam D. Barratt <adam@adam-barratt.org.uk>  Thu, 13 Mar 2008 19:33:38 +0000


scripts/debrsign.sh
#!/bin/bash

# This program is used to REMOTELY sign a .dsc and .changes file
# pair in the form needed for a legal Debian upload.  It is based on
# dpkg-buildpackage and debsign (which is also part of the devscripts
# package).
#
# In order for this program to work, debsign must be installed
# on the REMOTE machine which will be used to sign your package.
# You should run this program from within the package directory on
# the build machine.
#

# Debian GNU/Linux debrsign.
# Copyright 1999 Mike Goldman, all rights reserved
# Modifications copyright 1999 Julian Gilbey <jdg@debian.org>,
# all rights reserved.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <https://www.gnu.org/licenses/>.

# Abort if anything goes wrong
set -e

PROGNAME=`basename $0`

usage () {
    echo \
"Usage: debrsign [options] [username@]remotehost [changes or dsc]
  Options:
    -p<sign-command> The command to use for signing
    -e<maintainer>  Sign using key of <maintainer> (takes precedence over -m)
    -m<maintainer>  The same as -e
    -k<keyid>       The key to use for signing
    -S              Use changes file made for source-only upload
    -a<arch>        Use changes file made for Debian target architecture <arch>
    -t<target>      Use changes file made for GNU target architecture <target>
    --multi         Use most recent multiarch .changes file found
    --path          Specify directory GPG binary is located on remote host
    --help          Show this message
    --version       Show version and copyright information
  If a changes or dscfile is specified, it is signed, otherwise
  debian/changelog is parsed to find the changes file.  The signing
  is performed on remotehost using ssh and debsign."
}

version () {
    echo \
"This is debrsign, from the Debian devscripts package, version ###VERSION###
This code is copyright 1999 by Mike Goldman and Julian Gilbey,
all rights reserved.  This program comes with ABSOLUTELY NO WARRANTY.
You are free to redistribute this code under the terms of the
GNU General Public License, version 2 or later."
}

mustsetvar () {
    if [ "x$2" = x ]
    then
        echo >&2 "$PROGNAME: unable to determine $3"
        exit 1
    else
        # echo "$PROGNAME: $3 is $2"
        eval "$1=\"\$2\""
    fi
}

withecho () {
    echo " $@"
    "$@"
}

# --- main script

# For partial security, even though we know it doesn't work :(
# I guess maintainers will have to be careful, and there's no way around
# this in a shell script.
unset IFS
PATH=/usr/local/bin:/usr/bin:/bin
umask `perl -e 'printf "%03o\n", umask | 022'`

eval $(
    set +e
    for var in $VARS; do
        eval "$var=\$DEFAULT_$var"
    done
    for file in /etc/devscripts.conf ~/.devscripts; do
      [ -r $file ] && . $file
    done

    set | grep '^DEBRSIGN_')

signargs=
while [ $# != 0 ]
do
    value="`echo x\"$1\" | sed -e 's/^x-.//'`"
    case "$1" in
        -S)         sourceonly="true" ;;
        -a*)        targetarch="$value" ;;
        -t*)        targetgnusystem="$value" ;;
        --multi)    multiarch="true" ;;
        --help)     usage; exit 0 ;;
        --version)  version; exit 0 ;;
        --path)     DEBRSIGN_PGP_PATH="$value" ;;
        -*)         signargs="$signargs '$1'" ;;
        *)          break ;;
    esac
    shift
done

# Command line parameters are remote host (mandatory) and changes file
# name (optional).  If there is no changes file name, we must be at the
# top level of a source tree and will figure out its name from
# debian/changelog
case $# in
    2)  remotehost="$1"
        case "$2" in
            *.dsc)
                changes=
                dsc=$2
                ;;
            *.changes)
                changes=$2
                dsc=`echo $changes | \
                    perl -pe 's/\.changes$/.dsc/; s/(.*)_(.*)_(.*)\.dsc/\1_\2.dsc/'`
                buildinfo=`echo $changes | \
                    perl -pe 's/\.changes$/.buildinfo/; s/(.*)_(.*)_(.*)\.buildinfo/\1_\2_\3.buildinfo/'`
                ;;
            *)
                echo "$PROGNAME: Only a .changes or .dsc file is allowed as second argument!" >&2
                exit 1
                ;;
        esac
        ;;

    1)  remotehost="$1"
        case "$1" in
            *.changes)
                echo "$PROGNAME: You must pass the address of the signing host as as the first argument" >&2
                exit 1
                ;;
            *)
                # We have to parse debian/changelog to find the current version
                if [ ! -r debian/changelog ]; then
                    echo "$PROGNAME: Must be run from top of source dir or a .changes file given as arg" >&2
                    exit 1
                fi
                ;;
        esac


        mustsetvar package "`dpkg-parsechangelog -SSource`" "source package"
        mustsetvar version "`dpkg-parsechangelog -SVersion`" "source version"

        if [ "x$sourceonly" = x ]
        then
            if [ -n "$targetarch" ] && [ -n "$targetgnusystem" ]; then
                mustsetvar arch "$(dpkg-architecture "-a${targetarch}" "-t${targetgnusystem}" -qDEB_HOST_ARCH)" "build architecture"
            elif [ -n "$targetarch" ]; then
                mustsetvar arch "$(dpkg-architecture "-a${targetarch}" -qDEB_HOST_ARCH)" "build architecture"
            elif [ -n "$targetgnusystem" ]; then
                mustsetvar arch "$(dpkg-architecture "-t${targetgnusystem}" -qDEB_HOST_ARCH)" "build architecture"
            else
                mustsetvar arch "$(dpkg-architecture -qDEB_HOST_ARCH)" "build architecture"
            fi
        else
            arch=source
        fi

        sversion=`echo "$version" | perl -pe 's/^\d+://'`
        pv="${package}_${sversion}"
        pva="${package}_${sversion}${arch:+_${arch}}"
        dsc="../$pv.dsc"
        buildinfo="../$pva.buildinfo"
        changes="../$pva.changes"
        if [ -n "$multiarch" -o ! -r $changes ]; then
            changes=$(ls "../${package}_${sversion}_*+*.changes" "../${package}_${sversion}_multi.changes" 2>/dev/null | head -1)
            if [ -z "$multiarch" ]; then
                if [ -n "$changes" ]; then
                    echo "$PROGNAME: could not find normal .changes file but found multiarch file:" >&2
                    echo "  $changes" >&2
                    echo "Using this changes file instead." >&2
                else
                    echo "$PROGNAME: Can't find or can't read changes file $changes!" >&2
                    exit 1
                fi
            elif [ -n "$multiarch" -a -z "$changes" ]; then
                echo "$PROGNAME: could not find any multiarch .changes file with name" >&2
                echo "../${package}_${sversion}_*.changes" >&2
                exit 1
            fi
        fi
        ;;

    *)        echo "Usage: $PROGNAME [options] [user@]remotehost [.changes or .dsc file]" >&2
        exit 1 ;;
esac

if [ "x$remotehost" == "x" ]
then
        echo "No [user@]remotehost specified!" >&2
        exit 1
fi

declare -A base
base["$changes"]=`basename "$changes"`
base["$dsc"]=`basename "$dsc"`
base["$buildinfo"]=`basename "$buildinfo"`

if [ -n "$changes" ]
then
    if [ ! -f "$changes" -o ! -r "$changes" ]
    then
        echo "Can't find or can't read changes file $changes!" >&2
        exit 1
    fi

    # Is there a dsc file listed in the changes file?
    if grep -q "${base[$dsc]}" "$changes"
    then
        if [ ! -f "$dsc" -o ! -r "$dsc" ]
        then
            echo "Can't find or can't read dsc file $dsc!" >&2
            exit 1
        fi
    else
        unset base["$dsc"]
    fi
    # Is there a buildinfo file listed in the changes file?
    if grep -q "${base[$buildinfo]}" "$changes"
    then
        if [ ! -f "$buildinfo" -o ! -r "$buildinfo" ]
        then
            echo "Can't find or can't read buildinfo file $buildinfo!" >&2
            exit 1
        fi
    else
        unset base["$buildinfo"]
    fi
    # Now do the real work
    withecho scp "${!base[@]}" "$remotehost:\$HOME"
    withecho ssh -t "$remotehost" "debsign $signargs ${base[$changes]}"
    for file in "${!base[@]}"
    do
        withecho scp "$remotehost:\$HOME/${base["$file"]}" "$file"
    done
    withecho ssh "$remotehost" "rm -f ${base[@]}"

    echo "Successfully signed changes file"
else
    if [ ! -f "$dsc" -o ! -r "$dsc" ]
    then
        echo "Can't find or can't read dsc file $dsc!" >&2
        exit 1
    fi

    withecho scp "$dsc" "$remotehost:\$HOME"
    withecho ssh -t "$remotehost" "${DEBRSIGN_PGP_PATH}debsign $signargs $dscbase"
    withecho scp "$remotehost:\$HOME/$dscbase" "$dsc"
    withecho ssh "$remotehost" "rm -f $dscbase"

    echo "Successfully signed dsc file"
fi
exit 0
